### PR TITLE
Shiranr/anonymizer v2 engine

### PR DESCRIPTION
**Pull request by shiranr**

### CODE REVIEW

File: presidio-anonymizer/presidio_anonymizer/anonymizer_engine.py [hunk 1]
```diff
@@ -10,39 +16,38 @@ class AnonymizerEngine:
...
-        # TODO a loop that goes through the analyzer results from END to START! reverse.
-        # TODO for each result, replace the old value with the new value
-        # Make sure we handle partial intersections using the endpoint param
-        # TODO dictionary with the transformations and fields
-        # To map field to its transformation
-        pass
+        analyzer_results = self._analyze_results.to_sorted_set(True)
+        for result in analyzer_results:
+            transformation = self._transformations.get_transformation(result)
+            anonymizer_class = anonymizers.get(transformation.get("type").lower())
```
Comment (SharonHart) [inline presidio-anonymizer/presidio_anonymizer/anonymizer_engine.py:44]: I would have a map between the type and the class, less hacky, and will also raise a more specifc exception in case a non-existent type is passed

File: presidio-anonymizer/presidio_anonymizer/anonymizer_engine.py [hunk 2]
```diff
@@ -10,39 +16,38 @@ class AnonymizerEngine:
...
 
         :return: the anonymized text
         """
-        # TODO a loop that goes through the analyzer results from END to START! reverse.
-        # TODO for each result, replace the old value with the new value
-        # Make sure we handle partial intersections using the endpoint param
-        # TODO dictionary with the transformations and fields
-        # To map field to its transformation
-        pass
+        analyzer_results = self._analyze_results.to_sorted_set(True)
```
Comment (SharonHart) [inline presidio-anonymizer/presidio_anonymizer/anonymizer_engine.py:41]: Since we define it as a standalone service, me might want to use different terminology, and not mention 'analyzer' at all.

File: presidio-anonymizer/tests/test_aonnymizer_domain.py [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+    data = AnalyzerResult.validate_and_create(content)
+    assert data.start == content.get("start")
+    assert data.end == content.get("end")
+    assert data.score == content.get("score")
+    assert data.entity_type == content.get("entity_type")
+
+
+@pytest.mark.parametrize(
+    "request_json, result_text",
+    [
```
Comment (SharonHart) [inline presidio-anonymizer/tests/test_aonnymizer_domain.py:64]: Black autoformats parametrized tests its to a non readable form, you can have it ignore those wrapping in a fmt: on/off block, see what I did in the flake8 plugins PR

File: presidio-anonymizer/presidio_anonymizer/anonymizer_engine.py (hunk 1 above)
Comment (shiranr) [reply presidio-anonymizer/presidio_anonymizer/anonymizer_engine.py:44]: The mapping between the type and the class is a map.

Comment (SharonHart) [reply presidio-anonymizer/presidio_anonymizer/anonymizer_engine.py:44]: right, my bad.

File: presidio-anonymizer/presidio_anonymizer/anonymizers/replace.py [hunk 4]
```diff
@@ -1,15 +1,12 @@
...
-    def __init__(self,
-                 new_text: str):
-        self.new_text = new_text
-
-    def anonymize(self):
-        """:return: new_text."""
-        pass
+    def anonymize(self, params={}):
+        """:return: new_val."""
+        new_val = params.get("new_value")
```
Comment (itye-msft) [inline presidio-anonymizer/presidio_anonymizer/anonymizers/replace.py:11]: I suggest not to leave implementations like that.
If it's not implemented, it's better return nothing, or throw not implemented exception.
The doc string is also not relevant and should be removed for now.

File: presidio-anonymizer/presidio_anonymizer/entities/analyzer_result.py [hunk 5]
```diff
@@ -0,0 +1,101 @@
+"""
+AnalyzerResult is the exact copy of the recognizer result.
```
Comment (itye-msft) [inline presidio-anonymizer/presidio_anonymizer/entities/analyzer_result.py:2]: I don't think this line is required.

File: presidio-anonymizer/presidio_anonymizer/entities/analyzer_result.py [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+
+Represents the findings of detected entity.
+"""
+from presidio_anonymizer.entities.invalid_exception import InvalidParamException
+
+
+class AnalyzerResult(object):
+    """
+    AnalyzerResult is a duplication of the recognizer result.
```
Comment (itye-msft) [inline presidio-anonymizer/presidio_anonymizer/entities/analyzer_result.py:11]: This line is not required

File: presidio-anonymizer/presidio_anonymizer/entities/analyzer_result.py [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+
+    Suppose to validate and compare an recognizer result object.
```
Comment (itye-msft) [inline presidio-anonymizer/presidio_anonymizer/entities/analyzer_result.py:13]: I would suggest to leave out the word `Suppose`

File: presidio-anonymizer/presidio_anonymizer/entities/analyzer_result.py [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+            other) and self.entity_type.__eq__(
+            other.entity_type) and self.score is other.score
+
+    def __hash__(self):
+        """
+        Hash the result data by using all class fields.
+
+        :return: int
+        """
+        return hash(str(self.start) + " " + str(self.end) + " " + str(
```
Comment (itye-msft) [inline presidio-anonymizer/presidio_anonymizer/entities/analyzer_result.py:80]: Optional:
In terms of efficiency, string interpolation is faster than concatenation.
e.g. "{} {} {} {}".format(str(self.start), str(self.end), str(self.score), self.entity_type))
Or the inline interpolation as well.

File: presidio-anonymizer/presidio_anonymizer/entities/analyzer_results.py [hunk 9]
```diff
@@ -0,0 +1,63 @@
+"""Wrapper to the list of AnalyzerResult which manipulates the list."""
+from typing import List
+
+
+class AnalyzerResults(List):
+    """
+    Receives the analyzer result list and manipulates it.
```
Comment (itye-msft) [inline presidio-anonymizer/presidio_anonymizer/entities/analyzer_results.py:7]:
```suggestion
    A class which provides operations over the analyzer result list.
```

File: presidio-anonymizer/presidio_anonymizer/entities/analyzer_results.py [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+
+     The manipulation contains removal of unused results and sort by indices order.
```
Comment (itye-msft) [inline presidio-anonymizer/presidio_anonymizer/entities/analyzer_results.py:9]:
```suggestion
     For example:  removal of unused results and sort by indices order.
```

File: presidio-anonymizer/presidio_anonymizer/entities/analyzer_results.py [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+     According to the logic of:
```
Comment (itye-msft) [inline presidio-anonymizer/presidio_anonymizer/entities/analyzer_results.py:10]:
```suggestion
     Additional information about the rationale of this class:
```

File: presidio-anonymizer/presidio_anonymizer/entities/analyzer_results.py [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+    - Full overlap of PIIs - When one text have several PIIs, the PII with the higher
+    score will be taken.
+    Between PIIs with similar scores, the selection will be arbitrary.
+    - One PII is contained in another - anonymizer will use the PII with larger text.
+    - Partial intersection - both will be returned concatenated.
+    """
+
+    def to_sorted_set(self, reverse=False):
+        """
+        Manipulate the list.
```
Comment (itye-msft) [inline presidio-anonymizer/presidio_anonymizer/entities/analyzer_results.py:22]: This line is more confusing than helpful. I suggest to remove it.

File: presidio-anonymizer/presidio_anonymizer/entities/analyzer_results.py [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+
+        remove_dups - removes results which impact the same text and should be ignored.
```
Comment (itye-msft) [inline presidio-anonymizer/presidio_anonymizer/entities/analyzer_results.py:24]: I don't understand where all the flags are used?
For example, where to put `remove_dups`? is it an input somewhere? 
I was looking for it in the code and I don't understand it in the context of this function which has 2 lines of code.
As a contributor in the future of presidio, what should I do with this info? Is it the correct place for it? should be move to a different function?

File: presidio-anonymizer/presidio_anonymizer/entities/engine_request.py [hunk 14]
```diff
@@ -0,0 +1,98 @@
...
+from presidio_anonymizer.anonymizers.hash import Hash
+from presidio_anonymizer.anonymizers.mask import Mask
+from presidio_anonymizer.anonymizers.redact import Redact
+from presidio_anonymizer.anonymizers.replace import Replace
+from presidio_anonymizer.entities.analyzer_result import AnalyzerResult
+from presidio_anonymizer.entities.analyzer_results import AnalyzerResults
+from presidio_anonymizer.entities.invalid_exception import InvalidParamException
+
+
+class AnonymizerEngineRequest:
```
Comment (itye-msft) [inline presidio-anonymizer/presidio_anonymizer/entities/engine_request.py:18]:
```suggestion
class AnonymizerRequest:
```
And also reflect class name in the file name as well please

File: presidio-anonymizer/presidio_anonymizer/entities/engine_request.py [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+    """The request the engine will get and act on."""
```
Comment (itye-msft) [inline presidio-anonymizer/presidio_anonymizer/entities/engine_request.py:19]:
```suggestion
    """Input validation for the anonymize process"""
```
Isn't it more accurate about what is does?

File: presidio-anonymizer/presidio_anonymizer/entities/engine_request.py [hunk 16]
```diff
@@ -0,0 +1,98 @@
...
+
+    anonymizers = {"mask": Mask, "fpe": FPE, "replace": Replace, "hash": Hash,
+                   "redact": Redact}
+
+    logger = logging.getLogger("presidio-anonymizer")
+
+    def __init__(self,
+                 data: dict):
+        """Handle text replacement for PIIs with requested transformations.
```
Comment (itye-msft) [inline presidio-anonymizer/presidio_anonymizer/entities/engine_request.py:28]: This is a constructor, so I think this doc is more relevant to the class level doc.
And I also think that it's not "replacement" only.
I would just remove this line.

File: presidio-anonymizer/presidio_anonymizer/entities/engine_request.py [hunk 17]
```diff
@@ -0,0 +1,98 @@
...
+    def __handle_analyzer_results(self, data):
+        results = data.get("analyzer_results")
+        if results is None or len(results) == 0:
+            raise InvalidParamException("Invalid input, "
+                                        "analyzer results can not be empty")
+        for result in results:
+            self._analysis_results.append(
+                AnalyzerResult.validate_and_create(result))
+
+    def __handle_transformations(self, data):
```
Comment (itye-msft) [inline presidio-anonymizer/presidio_anonymizer/entities/engine_request.py:78]: It would be nice and helpful to provide a short doc string for those functions, with  complicated logic.

File: presidio-anonymizer/tests/test_annonymizer_engine.py [hunk 18]
```diff
@@ -0,0 +1,16 @@
+import json
+import os
+
+from presidio_anonymizer.entities import AnonymizerEngineRequest
+
+from presidio_anonymizer import AnonymizerEngine
+
+
+def test_engine_result():
```
Comment (itye-msft) [inline presidio-anonymizer/tests/test_annonymizer_engine.py:9]: I would recommend to follow the unit test naming conventions we usually follow in our projects, which is in the classic form of:
`test_given<scenario>_then<expected result>`

File: presidio-anonymizer/tests/test_presidio_app.py [hunk 19]
```diff
@@ -0,0 +1,13 @@
+import json
+import os
+
+import app
+
+
+def test_anonymize():
```
Comment (itye-msft) [inline presidio-anonymizer/tests/test_presidio_app.py:7]: This is an integration test. which is great. Should consider using the Integration test framework which @SharonHart   contributed
It should not run in the context of regular unit tests

**Review by itye-msft — CHANGES_REQUESTED** [review]
Great work @shiranr !
I enjoyed reviewing this one.
I think most of my remarks are about naming and docs.

File: presidio-anonymizer/presidio_anonymizer/anonymizer_engine.py [hunk 20]
```diff
@@ -9,40 +11,31 @@ class AnonymizerEngine:
...
-        :param analyze_results: The results of the analyzer of PIIs locations and scores
+        :param data: a map which contains the transformations, analyzer_results and text
         """
-        self.transformations = transformations
-        self.analyze_results = analyze_results
-        self.text = text
-        self._end_point = len(text)
 
-    def anonymize(self):
+    def anonymize(self, engine_request: AnonymizerEngineRequest):
```
[the remark this answers is not in the document]
Comment (shiranr) [reply presidio-anonymizer/presidio_anonymizer/anonymizer_engine.py:24]: DONE

File: presidio-anonymizer/presidio_anonymizer/anonymizer_engine.py [hunk 21]
```diff
@@ -9,40 +11,31 @@ class AnonymizerEngine:
...
-        # TODO for each result, replace the old value with the new value
-        # Make sure we handle partial intersections using the endpoint param
-        # TODO dictionary with the transformations and fields
-        # To map field to its transformation
-        pass
+        original_text = engine_request.get_text()
+        end_point = len(original_text)
+        text = engine_request.get_text()
+        analyzer_results = engine_request.get_analysis_results().to_sorted_set(True)
+        for result in analyzer_results:
```
[the remark this answers is not in the document]
Comment (shiranr) [reply presidio-anonymizer/presidio_anonymizer/anonymizer_engine.py:33]: DONE

File: presidio-anonymizer/presidio_anonymizer/anonymizer_engine.py [hunk 22]
```diff
@@ -9,40 +11,31 @@ class AnonymizerEngine:
...
         :return: the anonymized text
         """
-        # TODO a loop that goes through the analyzer results from END to START! reverse.
-        # TODO for each result, replace the old value with the new value
-        # Make sure we handle partial intersections using the endpoint param
-        # TODO dictionary with the transformations and fields
-        # To map field to its transformation
-        pass
+        original_text = engine_request.get_text()
+        end_point = len(original_text)
```
[the remark this answers is not in the document]
Comment (shiranr) [reply presidio-anonymizer/presidio_anonymizer/anonymizer_engine.py:30]: DONE.

File: presidio-anonymizer/presidio_anonymizer/anonymizer_engine.py [hunk 23]
```diff
@@ -9,40 +11,31 @@ class AnonymizerEngine:
...
+        text = engine_request.get_text()
+        analyzer_results = engine_request.get_analysis_results().to_sorted_set(True)
+        for result in analyzer_results:
+            transformation = engine_request.get_transformation(result)
+            self.logger.debug(f"received transformation {transformation.get('type')}")
+            anonymizer_class = transformation.get("anonymizer")
```
[the remark this answers is not in the document]
Comment (shiranr) [reply presidio-anonymizer/presidio_anonymizer/anonymizer_engine.py:36]: It's parallel to analyzer engine. 
We are keeping some kind of convention.
Also, this is where the "magic" happens.
Also, the anonymizers are currently not working (not implemented under this user story), there will be more things here once they do.

File: presidio-anonymizer/presidio_anonymizer/entities/engine_request.py (hunk 16 above)
Comment (shiranr) [reply presidio-anonymizer/presidio_anonymizer/entities/engine_request.py:28]: I can't remove this line :( linter...

File: presidio-anonymizer/presidio_anonymizer/entities/engine_request.py (hunk 17 above)
Comment (shiranr) [reply presidio-anonymizer/presidio_anonymizer/entities/engine_request.py:78]: DONE.

File: presidio-anonymizer/tests/test_annonymizer_engine.py (hunk 18 above)
Comment (shiranr) [reply presidio-anonymizer/tests/test_annonymizer_engine.py:9]: DONE.

File: presidio-anonymizer/tests/test_presidio_app.py (hunk 19 above)
Comment (shiranr) [reply presidio-anonymizer/tests/test_presidio_app.py:7]: I will do it. But will create a new ticket for it since I need to close some other things first :)

Comment (shiranr) [reply presidio-anonymizer/tests/test_presidio_app.py:7]: Ticket 2664

File: presidio-anonymizer/presidio_anonymizer/entities/engine_request.py (hunk 15 above)
Comment (shiranr) [reply presidio-anonymizer/presidio_anonymizer/entities/engine_request.py:19]: DONE.

File: presidio-anonymizer/presidio_anonymizer/entities/engine_request.py (hunk 14 above)
Comment (shiranr) [reply presidio-anonymizer/presidio_anonymizer/entities/engine_request.py:18]: DONE

File: presidio-anonymizer/presidio_anonymizer/entities/analyzer_results.py (hunk 13 above)
Comment (shiranr) [reply presidio-anonymizer/presidio_anonymizer/entities/analyzer_results.py:24]: DONE

File: presidio-anonymizer/presidio_anonymizer/entities/analyzer_results.py (hunk 12 above)
Comment (shiranr) [reply presidio-anonymizer/presidio_anonymizer/entities/analyzer_results.py:22]: Can't. linter. changed it.

File: presidio-anonymizer/presidio_anonymizer/entities/analyzer_results.py (hunk 10 above)
Comment (shiranr) [reply presidio-anonymizer/presidio_anonymizer/entities/analyzer_results.py:9]: it's not an example. that's what it contains.

File: presidio-anonymizer/presidio_anonymizer/entities/analyzer_results.py (hunk 9 above)
Comment (shiranr) [reply presidio-anonymizer/presidio_anonymizer/entities/analyzer_results.py:7]: Done.

File: presidio-anonymizer/presidio_anonymizer/entities/analyzer_result.py (hunk 8 above)
Comment (shiranr) [reply presidio-anonymizer/presidio_anonymizer/entities/analyzer_result.py:80]: DONE

File: presidio-anonymizer/presidio_anonymizer/entities/analyzer_result.py (hunk 7 above)
Comment (shiranr) [reply presidio-anonymizer/presidio_anonymizer/entities/analyzer_result.py:13]: Done

File: presidio-anonymizer/presidio_anonymizer/entities/analyzer_result.py (hunk 6 above)
Comment (shiranr) [reply presidio-anonymizer/presidio_anonymizer/entities/analyzer_result.py:11]: Sadly, the linter thinks otherwise.

File: presidio-anonymizer/presidio_anonymizer/entities/analyzer_result.py (hunk 5 above)
Comment (shiranr) [reply presidio-anonymizer/presidio_anonymizer/entities/analyzer_result.py:2]: Sadly, the linter thinks otherwise.

File: presidio-anonymizer/presidio_anonymizer/anonymizers/replace.py (hunk 4 above)
Comment (shiranr) [reply presidio-anonymizer/presidio_anonymizer/anonymizers/replace.py:11]: Each class like this is marked with a TODO of test and implement. 
I know usually we need to put an exception but I need this to pass the tests otherwise the flow won't work.
Also, there are tickets for it as well

File: presidio-anonymizer/presidio_anonymizer/anonymizers/fpe.py [hunk 24]
```diff
@@ -14,16 +14,10 @@ class FPE(Anonymizer):
...
-
-    def anonymize(self):
+    def anonymize(self, params={}):
         """Return anonymized text using FF1 algorithm."""
-        pass
+        old_text = params.get("old_text")
+        decrypt = params.get("decrypt")
+        tweak = params.get("tweak")
+        key = params.get("key")
+        return old_text + decrypt + tweak + key
```
[the remark this answers is not in the document]
Comment (shiranr) [reply presidio-anonymizer/presidio_anonymizer/anonymizers/fpe.py:23]: This is not a implementation. Same as the remark below:
Each class like this is marked with a TODO of test and implement.
I know usually we need to put an exception but I need this to pass the tests otherwise the flow won't work.
Also, there are tickets for it as well

File: presidio-anonymizer/presidio_anonymizer/anonymizers/mask.py [hunk 25]
```diff
@@ -1,25 +1,19 @@
...
-
-    def anonymize(self):
+    def anonymize(self, params={}):
         """
         Anonymize a given amount of text with a given character.
 
         :return: The given text masked as requested
         """
-        pass
+        # chars_to_replace = params.get("chars_to_replace")
```
[the remark this answers is not in the document]
Comment (shiranr) [reply presidio-anonymizer/presidio_anonymizer/anonymizers/mask.py:15]: I need this for the implementation to know which params need to be there. 
I don't want to go looking for them again in the code. 
This needs to be implemented.

File: presidio-anonymizer/presidio_anonymizer/anonymizers/hash.py [hunk 26]
```diff
@@ -5,14 +5,10 @@ class Hash(object):
...
-
-    def anonymize(self):
+    def anonymize(self, params={}):
         """
         Hash given value using sha256.
 
         :return: hashed original text
         """
-        pass
+        return params.get("old_text")
```
[the remark this answers is not in the document]
Comment (shiranr) [reply presidio-anonymizer/presidio_anonymizer/anonymizers/hash.py:14]: This will be implemented please see remarks below

File: presidio-anonymizer/tests/test_aonnymizer_domain.py (hunk 3 above)
Comment (shiranr) [reply presidio-anonymizer/tests/test_aonnymizer_domain.py]: Done.

Comment (shiranr) [reply presidio-anonymizer/tests/test_aonnymizer_domain.py]: Done.

File: presidio-anonymizer/presidio_anonymizer/anonymizer_engine.py (hunk 2 above)
Comment (shiranr) [reply presidio-anonymizer/presidio_anonymizer/anonymizer_engine.py]: Well an analyzer is an anlalyzer not necessarily ours.
It's the results from the analyzer :)
So I don't think it makes  a difference.

Comment (SharonHart) [reply presidio-anonymizer/presidio_anonymizer/anonymizer_engine.py]: I mean it can just mask data, not result from other service, not analyzation service... these kind of stuff start a conceptual coupling that lead to code coupling, just my 2 cents

Comment (shiranr) [reply presidio-anonymizer/presidio_anonymizer/anonymizer_engine.py]: You are correct... 
But I think it's a great change, all around. Lets consult on this one with the team and open a use story accordingly wdyt?

File: presidio-anonymizer/app.py [hunk 27]
```diff
@@ -1,21 +1,27 @@
-"""TODO: Flask application instantiation and route definition."""
-
 from flask import Flask, request
 
+from presidio_anonymizer.anonymizer_engine import AnonymizerEngine
```
[the remark this answers is not in the document]
Comment (shiranr) [reply presidio-anonymizer/app.py:5]: Done

File: presidio-anonymizer/presidio_anonymizer/anonymizer_engine.py (hunk 2 above)
Comment (SharonHart) [reply presidio-anonymizer/presidio_anonymizer/anonymizer_engine.py]: Yes definitely, it's a product decision even more then engineering.
We tackled the same issue in the REST API ADR, we can continue there with product

File: presidio-anonymizer/presidio_anonymizer/anonymizer_engine.py (hunk 23 above)
[the remark this answers is not in the document]
Comment (itye-msft) [reply presidio-anonymizer/presidio_anonymizer/anonymizer_engine.py:37]: "We are keeping some kind of convention." **Where?**

File: presidio-anonymizer/presidio_anonymizer/anonymizers/fpe.py (hunk 24 above)
[the remark this answers is not in the document]
Comment (itye-msft) [reply presidio-anonymizer/presidio_anonymizer/anonymizers/fpe.py:23]: How can tests pass if it's not implemented?
Tests should fail, or just don't add them until implemented.
It's better return "" than to push this logic to the repo

File: presidio-anonymizer/presidio_anonymizer/anonymizers/hash.py (hunk 26 above)
[the remark this answers is not in the document]
Comment (itye-msft) [reply presidio-anonymizer/presidio_anonymizer/anonymizers/hash.py:14]: Again, if this is not implemented, don't return this.
Throw not implemented exception.

File: presidio-anonymizer/presidio_anonymizer/anonymizers/mask.py (hunk 25 above)
[the remark this answers is not in the document]
Comment (itye-msft) [reply presidio-anonymizer/presidio_anonymizer/anonymizers/mask.py:15]: As a reviewer, how should I know that?
It's better communicate this with a comment, why it's commented, and a user story ID.

File: presidio-anonymizer/presidio_anonymizer/anonymizers/replace.py (hunk 4 above)
Comment (itye-msft) [reply presidio-anonymizer/presidio_anonymizer/anonymizers/replace.py:11]: Please see the last line of my comment about the doc string.

File: presidio-anonymizer/presidio_anonymizer/entities/analyzer_result.py (hunk 5 above)
Comment (itye-msft) [reply presidio-anonymizer/presidio_anonymizer/entities/analyzer_result.py:2]: Wdym? The whole comment - where does it belong to?
I think you can just remove the whole comment block

File: presidio-anonymizer/presidio_anonymizer/entities/analyzer_result.py (hunk 6 above)
Comment (itye-msft) [reply presidio-anonymizer/presidio_anonymizer/entities/analyzer_result.py:11]: Let me rephrase: The text in this docstring is unclear.
```suggestion
    AnalyzerResult is the output of the analyze process.
```

File: presidio-anonymizer/presidio_anonymizer/entities/analyzer_results.py (hunk 10 above)
Comment (itye-msft) [reply presidio-anonymizer/presidio_anonymizer/entities/analyzer_results.py:9]: Let me rephrase: The word manipulation is not a good fit here. You can say "It includes ...".
In general it's hard to read this documentation

**Review by itye-msft — CHANGES_REQUESTED** [review]

**Review by itye-msft — APPROVED** [review]

**Review by SharonHart — APPROVED** [review]

**Review by omri374 — APPROVED** [review]
Great work!

File: presidio-anonymizer/app.py [hunk 28]
```diff
@@ -1,21 +1,38 @@
...
+                # TODO add logger 2652
+                return "Internal server error", 500
+            return text
 
 
 if __name__ == "__main__":
-    app.run()
+    port = int(os.environ.get("PORT", DEFAULT_PORT))
+    server = Server()
+    server.app.run(port=port)
```
Comment (SharonHart) [inline presidio-anonymizer/app.py:38]: This way, you cant pass the host from outside with --host, it will break running a dockerizerd app since it needs to override 127.0.0.1 to 0.0.0.0
@itye-msft

**Review by SharonHart — APPROVED** [review]

File: presidio-anonymizer/presidio_anonymizer/entities/analyzer_results.py [hunk 29]
```diff
@@ -0,0 +1,59 @@
...
+    - One PII - uses a given or default transformation to anonymize and replace the PII
+    text entity.
+    - Full overlap of PIIs - When one text have several PIIs, the PII with the higher
+    score will be taken.
+    Between PIIs with similar scores, the selection will be arbitrary.
+    - One PII is contained in another - anonymizer will use the PII with larger text.
+    - Partial intersection - both will be returned concatenated.
+    """
+
+    def to_sorted_unique_results(self, reverse=False) -> List[AnalyzerResult]:
```
Comment (SharonHart) [inline presidio-anonymizer/presidio_anonymizer/entities/analyzer_results.py:22]: I don't see why letting to reverse, and having false as default when the only usage is for reverse=True

File: presidio-anonymizer/presidio_anonymizer/entities/anonymizer_request.py [hunk 30]
```diff
@@ -0,0 +1,117 @@
...
+    def __handle_transformations(self, data):
+        """
+        Go over the transformations and get the relevant anonymizer class for it.
+
+        Inserts the class to the transformation so the engine will use it.
+        :param data: contains the text, transformations and analyzer_results
+        :return: None
+        """
+        transformations = data.get("transformations")
+        if transformations is not None:
```
Comment (SharonHart) [inline presidio-anonymizer/presidio_anonymizer/entities/anonymizer_request.py:93]: if transformations:
